File: firestore-https-cplus/firestore_client.cpp
#include "firestore_client.hpp"
#include "logger.hpp"

#include "httplib.h"

#include <algorithm>
#include <array>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <memory>
#include <random>
#include <sstream>
#include <stdexcept>

using nlohmann::json;

namespace {
// Helper to run shell commands
std::string exec(const char *cmd) {
  std::array<char, 128> buffer;
  std::string result;
  std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
  if (!pipe) {
    return "";
  }
  while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
    result += buffer.data();
  }
  // Trim newline
  while (!result.empty() && (result.back() == '\n' || result.back() == '\r')) {
    result.pop_back();
  }
  return result;
}

// Date/Time helpers
std::string time_point_to_iso(std::chrono::system_clock::time_point tp) {
  auto in_time_t = std::chrono::system_clock::to_time_t(tp);
  std::stringstream ss;
  ss << std::put_time(std::gmtime(&in_time_t), "%Y-%m-%dT%H:%M:%SZ");
  return ss.str();
}
} // namespace

FirestoreClient::FirestoreClient() {}

bool FirestoreClient::initialize() {
  try {
    log_json("INFO", "FirestoreClient::initialize(): calling get_project_id()");
    project_id_ = get_project_id();
    log_json("INFO", "project_id_", {{"project_id", project_id_}});
    if (project_id_.empty()) {
      log_json("ERROR", "GOOGLE_CLOUD_PROJECT environment variable not set.");
      return false;
    }

    log_json("INFO",
             "FirestoreClient::initialize(): calling get_access_token()");
    access_token_ = get_access_token();
    log_json("INFO", "access_token_ length",
             {{"length", (long)access_token_.length()}});
    if (access_token_.empty()) {
      log_json("ERROR", "Failed to obtain access token.");
      return false;
    }

    // Test connection
    log_json("INFO", "FirestoreClient::initialize(): testing connection to "
                     "firestore.googleapis.com");
    httplib::Client cli("https://firestore.googleapis.com");
    cli.set_bearer_token_auth(access_token_);
    std::string path = "/v1/projects/" + project_id_ +
                       "/databases/(default)/documents/inventory?pageSize=1";

    log_json("INFO", "FirestoreClient::initialize(): calling cli.Get",
             {{"path", path}});
    auto res = cli.Get(path.c_str());
    log_json("INFO", "FirestoreClient::initialize(): cli.Get returned");
    if (res && res->status == 200) {
      initialized_ = true;
      return true;
    } else {
      std::string status = res ? std::to_string(res->status) : "Unknown";
      std::string body = res ? res->body : "";
      log_json("ERROR", "Error connecting to Firestore",
               {{"status", status}, {"body", body}});
      return false;
    }
  } catch (const std::exception &e) {
    log_json("ERROR", "Exception initializing FirestoreClient",
             {{"error", e.what()}});
    return false;
  }
}

std::string FirestoreClient::get_project_id() {
  const char *env_p = std::getenv("GOOGLE_CLOUD_PROJECT");
  if (env_p)
    return std::string(env_p);

  // Try gcloud config
  return exec("gcloud config get-value project 2>/dev/null");
}

std::string FirestoreClient::get_access_token() {
  // Try Metadata Server first (Cloud Run / GCE)
  // Using a short timeout
  try {
    log_json("INFO",
             "get_access_token(): creating httplib::Client for metadata");
    httplib::Client cli("http://metadata.google.internal");
    log_json("INFO", "get_access_token(): setting timeouts");
    cli.set_connection_timeout(1); // 1 second timeout
    cli.set_read_timeout(1);

    httplib::Headers headers = {{"Metadata-Flavor", "Google"}};
    log_json("INFO", "get_access_token(): calling cli.Get() for metadata");
    auto res = cli.Get(
        "/computeMetadata/v1/instance/service-accounts/default/token", headers);
    log_json("INFO", "get_access_token(): cli.Get() returned");

    if (res && res->status == 200) {
      try {
        auto j = json::parse(res->body);
        return j["access_token"];
      } catch (...) {
      }
    }
  } catch (const std::exception &e) {
    log_json("WARNING", "get_access_token(): exception during metadata check",
             {{"error", e.what()}});
  } catch (...) {
    log_json("WARNING",
             "get_access_token(): unknown exception during metadata check");
  }

  // Fallback to gcloud
  log_json(
      "INFO",
      "get_access_token(): falling back to gcloud auth print-access-token");
  return exec("gcloud auth print-access-token 2>/dev/null");
}

void FirestoreClient::ensure_authenticated() {
  // Simple refresh if token is empty, in a real app check expiration
  if (access_token_.empty()) {
    access_token_ = get_access_token();
  }
}

json FirestoreClient::make_firestore_request(const std::string &method,
                                             const std::string &path,
                                             const json &body) {
  ensure_authenticated();
  httplib::Client cli("https://firestore.googleapis.com");
  cli.set_bearer_token_auth(access_token_);
  cli.set_connection_timeout(10);
  cli.set_read_timeout(30);

  std::string full_path =
      "/v1/projects/" + project_id_ + "/databases/(default)/documents" + path;

  httplib::Result res;
  if (method == "GET") {
    res = cli.Get(full_path.c_str());
  } else if (method == "POST") {
    res = cli.Post(full_path.c_str(), body.dump(), "application/json");
  } else if (method == "PATCH") {
    res = cli.Patch(full_path.c_str(), body.dump(), "application/json");
  } else if (method == "DELETE") {
    res = cli.Delete(full_path.c_str());
  }

  if (res && (res->status >= 200 && res->status < 300)) {
    if (res->body.empty())
      return json({});
    return json::parse(res->body);
  } else {
    std::string status = res ? std::to_string(res->status) : "null";
    std::string body = res ? res->body : "";
    log_json("ERROR", "Firestore Request Failed",
             {{"method", method},
              {"path", full_path},
              {"status", status},
              {"body", body}});
    throw std::runtime_error("Firestore request failed");
  }
}

json FirestoreClient::doc_to_product(const json &doc) {
  json product;

  // Extract ID from name "projects/.../documents/inventory/ID"
  std::string full_name = doc["name"];
  auto pos = full_name.find_last_of('/');
  if (pos != std::string::npos) {
    product["id"] = full_name.substr(pos + 1);
  }

  const auto &fields = doc["fields"];

  if (fields.contains("name"))
    product["name"] = fields["name"]["stringValue"];

  if (fields.contains("price")) {
    if (fields["price"].contains("integerValue"))
      product["price"] =
          std::stod(fields["price"]["integerValue"].get<std::string>());
    else if (fields["price"].contains("doubleValue"))
      product["price"] = fields["price"]["doubleValue"];
  }

  if (fields.contains("quantity")) {
    if (fields["quantity"].contains("integerValue"))
      product["quantity"] =
          std::stoi(fields["quantity"]["integerValue"].get<std::string>());
  }

  if (fields.contains("imgfile"))
    product["imgfile"] = fields["imgfile"]["stringValue"];

  if (fields.contains("timestamp"))
    product["timestamp"] = fields["timestamp"]["timestampValue"];
  if (fields.contains("actualdateadded"))
    product["actualdateadded"] = fields["actualdateadded"]["timestampValue"];

  return product;
}

json FirestoreClient::product_to_doc(const json &product) {
  json doc;
  json fields;

  if (product.contains("name"))
    fields["name"] = {{"stringValue", product["name"]}};

  if (product.contains("price")) {
    // Firestore requires doubleValue for floats or integerValue (as string) for
    // ints We'll use doubleValue for price to be safe, or integer if it's whole
    fields["price"] = {{"doubleValue", product["price"]}};
  }

  if (product.contains("quantity")) {
    fields["quantity"] = {
        {"integerValue", std::to_string(product["quantity"].get<int>())}};
  }

  if (product.contains("imgfile"))
    fields["imgfile"] = {{"stringValue", product["imgfile"]}};

  if (product.contains("timestamp"))
    fields["timestamp"] = {{"timestampValue", product["timestamp"]}};
  if (product.contains("actualdateadded"))
    fields["actualdateadded"] = {
        {"timestampValue", product["actualdateadded"]}};

  doc["fields"] = fields;
  return doc;
}

json FirestoreClient::get_products() {
  try {
    json response = make_firestore_request("GET", "/inventory");
    json products = json::array();

    if (response.contains("documents")) {
      for (const auto &doc : response["documents"]) {
        products.push_back(doc_to_product(doc));
      }
    }
    return products;
  } catch (...) {
    return json::array();
  }
}

json FirestoreClient::get_product(const std::string &id) {
  try {
    json doc = make_firestore_request("GET", "/inventory/" + id);
    return doc_to_product(doc);
  } catch (...) {
    return nullptr;
  }
}

json FirestoreClient::search_products(const std::string &query) {
  json all_products = get_products();
  json results = json::array();

  std::string q_lower = query;
  std::transform(q_lower.begin(), q_lower.end(), q_lower.begin(), ::tolower);

  for (const auto &p : all_products) {
    std::string name = p["name"];
    std::transform(name.begin(), name.end(), name.begin(), ::tolower);

    if (name.find(q_lower) != std::string::npos) {
      results.push_back(p);
    }
  }
  return results;
}

void FirestoreClient::add_or_update_product(const json &product) {
  // Check if exists by name
  json query_body = {{"structuredQuery",
                      {{"from", {{{"collectionId", "inventory"}}}},
                       {"where",
                        {{"fieldFilter",
                          {{"field", {{"fieldPath", "name"}}},
                           {"op", "EQUAL"},
                           {"value", {{"stringValue", product["name"]}}}}}}},
                       {"limit", 1}}}};

  // Using runQuery
  httplib::Client cli("https://firestore.googleapis.com");
  cli.set_bearer_token_auth(access_token_);
  std::string path =
      "/v1/projects/" + project_id_ + "/databases/(default)/documents:runQuery";
  auto res = cli.Post(path.c_str(), query_body.dump(), "application/json");

  std::string doc_id;
  if (res && res->status == 200) {
    json response = json::parse(res->body);
    // Response is array of results
    if (response.is_array() && !response.empty() &&
        response[0].contains("document")) {
      // Found existing
      std::string full_name = response[0]["document"]["name"];
      auto pos = full_name.find_last_of('/');
      doc_id = full_name.substr(pos + 1);
    }
  }

  json doc = product_to_doc(product);

  if (!doc_id.empty()) {
    // Update
    // Use PATCH with updateMask? Or just overwrite. TS uses update.
    // To keep it simple, we'll patch specific fields or just overwrite.
    // For REST, PATCHing /documents/inventory/{id} updates the document.
    make_firestore_request("PATCH", "/inventory/" + doc_id, doc);
  } else {
    // Create
    make_firestore_request("POST", "/inventory", doc);
  }
}

void FirestoreClient::reset() {
  // Get all and delete
  json all = get_products();
  for (const auto &p : all) {
    if (p.contains("id")) {
      try {
        make_firestore_request("DELETE",
                               "/inventory/" + p["id"].get<std::string>());
      } catch (...) {
      }
    }
  }
}

void FirestoreClient::seed() {
  std::vector<std::string> old_products = {
      "Apples",       "Bananas",         "Milk",          "Whole Wheat Bread",
      "Eggs",         "Cheddar Cheese",  "Whole Chicken", "Rice",
      "Black Beans",  "Bottled Water",   "Apple Juice",   "Cola",
      "Coffee Beans", "Green Tea",       "Watermelon",    "Broccoli",
      "Jasmine Rice", "Yogurt",          "Beef",          "Shrimp",
      "Walnuts",      "Sunflower Seeds", "Fresh Basil",   "Cinnamon"};

  std::random_device rd;
  std::mt19937 gen(rd());

  auto now = std::chrono::system_clock::now();

  for (const auto &name : old_products) {
    std::uniform_int_distribution<> price_dist(1, 10);
    std::uniform_int_distribution<> qty_dist(1, 500);
    std::uniform_real_distribution<> time_dist(
        0, 31536000000.0 / 1000.0); // approx year in seconds

    json p;
    p["name"] = name;
    p["price"] = price_dist(gen);
    p["quantity"] = qty_dist(gen);

    std::string img = "product-images/";
    std::string lower_name = name;
    std::transform(lower_name.begin(), lower_name.end(), lower_name.begin(),
                   ::tolower);
    // remove spaces
    lower_name.erase(std::remove(lower_name.begin(), lower_name.end(), ' '),
                     lower_name.end());
    p["imgfile"] = img + lower_name + ".png";

    auto past_time = now - std::chrono::seconds((long)time_dist(gen));
    p["timestamp"] = time_point_to_iso(past_time);
    p["actualdateadded"] = time_point_to_iso(now);

    add_or_update_product(p);
  }

  // Add "Recent" and "Out of Stock" logic similarly if needed, for brevity I'll
  // stick to this or expand if needed. The prompt asked to match TS, so I
  // should probably add the others.

  std::vector<std::string> recent = {"Parmesan Crisps",
                                     "Pineapple Kombucha",
                                     "Maple Almond Butter",
                                     "Mint Chocolate Cookies",
                                     "White Chocolate Caramel Corn",
                                     "Acai Smoothie Packs",
                                     "Smores Cereal",
                                     "Peanut Butter and Jelly Cups"};

  for (const auto &name : recent) {
    std::uniform_int_distribution<> price_dist(1, 10);
    std::uniform_int_distribution<> qty_dist(1, 100);

    json p;
    p["name"] = name;
    p["price"] = price_dist(gen);
    p["quantity"] = qty_dist(gen);
    std::string img = "product-images/";
    std::string lower_name = name;
    std::transform(lower_name.begin(), lower_name.end(), lower_name.begin(),
                   ::tolower);
    lower_name.erase(std::remove(lower_name.begin(), lower_name.end(), ' '),
                     lower_name.end());
    p["imgfile"] = img + lower_name + ".png";

    p["timestamp"] = time_point_to_iso(now); // recent
    p["actualdateadded"] = time_point_to_iso(now);

    add_or_update_product(p);
  }
}

File: firestore-stdio-swift/Sources/firestore-stdio-swift/FirestoreClient.swift
import AsyncHTTPClient
import Foundation
import Logging
import NIO
import NIOFoundationCompat
import NIOHTTP1

enum FirestoreError: Error {
  case invalidUrl
  case apiError(status: HTTPResponseStatus, message: String)
  case decodingError(Error)
  case missingData
}

actor FirestoreClient {
  private let auth: TokenProvider?
  private let httpClient: HTTPClient
  private let logger: Logger

  private func getBaseUrl() async throws -> String {
    guard let auth = auth else {
      throw FirestoreError.apiError(
        status: .internalServerError, message: "Firestore not initialized (missing authentication)")
    }
    let projectId = try await auth.getProjectId()
    return "https://firestore.googleapis.com/v1/projects/\(projectId)/databases/(default)/documents"
  }

  init(auth: TokenProvider?, httpClient: HTTPClient, logger: Logger) {
    self.auth = auth
    self.httpClient = httpClient
    self.logger = logger
  }

  private func getHeaders() async throws -> HTTPHeaders {
    guard let auth = auth else {
      throw FirestoreError.apiError(
        status: .internalServerError, message: "Firestore not initialized (missing authentication)")
    }
    let token = try await auth.getAccessToken()
    var headers = HTTPHeaders()
    headers.add(name: "Authorization", value: "Bearer \(token)")
    headers.add(name: "Content-Type", value: "application/json")
    return headers
  }

  // MARK: - CRUD

  func listProducts() async throws -> [Product] {
    let baseUrl = try await getBaseUrl()
    let url = "\(baseUrl)/inventory"
    var request = HTTPClientRequest(url: url)
    request.method = .GET
    request.headers = try await getHeaders()

    let response = try await httpClient.execute(request, timeout: .seconds(30))

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }

    let body = try await response.body.collect(upTo: 10 * 1024 * 1024)

    struct ListResponse: Codable {
      let documents: [FirestoreDocument]?
    }

    let listResponse = try JSONDecoder().decode(ListResponse.self, from: Data(buffer: body))
    return listResponse.documents?.compactMap { Product.from(document: $0) } ?? []
  }

  func getProduct(id: String) async throws -> Product? {
    let baseUrl = try await getBaseUrl()
    let url = "\(baseUrl)/inventory/\(id)"
    var request = HTTPClientRequest(url: url)
    request.method = .GET
    request.headers = try await getHeaders()

    let response = try await httpClient.execute(request, timeout: .seconds(10))

    if response.status == .notFound {
      return nil
    }

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }

    let body = try await response.body.collect(upTo: 1024 * 1024)
    let doc = try JSONDecoder().decode(FirestoreDocument.self, from: Data(buffer: body))
    return Product.from(document: doc)
  }

  func addProduct(_ product: Product) async throws {
    let baseUrl = try await getBaseUrl()
    let url = "\(baseUrl)/inventory"
    var request = HTTPClientRequest(url: url)
    request.method = .POST
    request.headers = try await getHeaders()

    let doc = FirestoreDocument(fields: product.toFirestoreFields())
    let bodyData = try JSONEncoder().encode(doc)
    request.body = .bytes(ByteBuffer(bytes: Array(bodyData)))

    let response = try await httpClient.execute(request, timeout: .seconds(10))

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }
  }

  func updateProduct(id: String, product: Product) async throws {
    let baseUrl = try await getBaseUrl()
    let url = "\(baseUrl)/inventory/\(id)"
    var request = HTTPClientRequest(url: url)
    request.method = .PATCH
    request.headers = try await getHeaders()

    let doc = FirestoreDocument(fields: product.toFirestoreFields())
    let bodyData = try JSONEncoder().encode(doc)
    request.body = .bytes(ByteBuffer(bytes: Array(bodyData)))

    let response = try await httpClient.execute(request, timeout: .seconds(10))

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }
  }

  func deleteProduct(id: String) async throws {
    let baseUrl = try await getBaseUrl()
    let url = "\(baseUrl)/inventory/\(id)"
    var request = HTTPClientRequest(url: url)
    request.method = .DELETE
    request.headers = try await getHeaders()

    let response = try await httpClient.execute(request, timeout: .seconds(10))

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }
  }

  // MARK: - Query

  func findProducts(name: String) async throws -> [Product] {
    let baseUrl = try await getBaseUrl()
    let url = "\(baseUrl):runQuery"
    var request = HTTPClientRequest(url: url)
    request.method = .POST
    request.headers = try await getHeaders()

    // Construct StructuredQuery
    let query: [String: Any] = [
      "structuredQuery": [
        "from": [["collectionId": "inventory"]],
        "where": [
          "fieldFilter": [
            "field": ["fieldPath": "name"],
            "op": "EQUAL",
            "value": ["stringValue": name],
          ]
        ],
      ]
    ]

    let bodyData = try JSONSerialization.data(withJSONObject: query)
    request.body = .bytes(ByteBuffer(bytes: Array(bodyData)))

    let response = try await httpClient.execute(request, timeout: .seconds(10))

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }

    let body = try await response.body.collect(upTo: 10 * 1024 * 1024)

    struct QueryResponseItem: Codable {
      let document: FirestoreDocument?
      let readTime: String?
    }

    let items = try JSONDecoder().decode([QueryResponseItem].self, from: Data(buffer: body))
    return items.compactMap { $0.document }.compactMap { Product.from(document: $0) }
  }

  // MARK: - Batch

  func batchDelete(ids: [String]) async throws {
    if ids.isEmpty { return }

    guard let auth = auth else {
      throw FirestoreError.apiError(
        status: .internalServerError, message: "Firestore not initialized (missing authentication)")
    }
    let projectId = try await auth.getProjectId()
    let url =
      "https://firestore.googleapis.com/v1/projects/\(projectId)/databases/(default)/documents:commit"
    var request = HTTPClientRequest(url: url)
    request.method = .POST
    request.headers = try await getHeaders()

    let writes = ids.map { id in
      return [
        "delete": "projects/\(projectId)/databases/(default)/documents/inventory/\(id)"
      ]
    }

    let body: [String: Any] = ["writes": writes]
    let bodyData = try JSONSerialization.data(withJSONObject: body)
    request.body = .bytes(ByteBuffer(bytes: Array(bodyData)))

    let response = try await httpClient.execute(request, timeout: .seconds(30))

    guard response.status == .ok else {
      let body = try await response.body.collect(upTo: 1024 * 1024)
      throw FirestoreError.apiError(status: response.status, message: String(buffer: body))
    }
  }
}
